vm/src/obj/objfloat.rs

```diff
@@ -179,132 +187,101 @@ impl PyFloat {
-            } else {
-                // certainly not equal, forward to inner_lt_int
-                inner_lt_int(self.value, other_int)
-            };
-
-            vm.ctx.new_bool(result)
-        } else {
-            vm.ctx.not_implemented()
-        }
+    fn le(&self, other: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        self.cmp(
+            other,
+            |a, b| a <= b,
+            |a, b| {
+                if let (Some(a_int), Some(b_float)) = (a.to_bigint(), b.to_f64()) {
+                    a <= b_float && a_int <= *b
+                } else {
+                    inner_lt_int(a, b)
+                }
+            },
+            vm,
+        )
     }
 
     #[pymethod(name = "__gt__")]
-    fn gt(&self, i2: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
-        let v1 = self.value;
-        if objtype::isinstance(&i2, &vm.ctx.float_type()) {
-            vm.ctx.new_bool(v1 > get_value(&i2))
-        } else if objtype::isinstance(&i2, &vm.ctx.int_type()) {
-            let other_int = objint::get_value(&i2);
-
-            vm.ctx.new_bool(inner_gt_int(self.value, other_int))
-        } else {
-            vm.ctx.not_implemented()
-        }
+    fn gt(&self, other: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        self.cmp(other, |a, b| a > b, |a, b| inner_gt_int(a, b), vm)
     }
 
     #[pymethod(name = "__ge__")]
-    fn ge(&self, i2: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
-        let v1 = self.value;
-        if objtype::isinstance(&i2, &vm.ctx.float_type()) {
-            vm.ctx.new_bool(v1 >= get_value(&i2))
-        } else if objtype::isinstance(&i2, &vm.ctx.int_type()) {
-            let other_int = objint::get_value(&i2);
-
-            let result = if let (Some(self_int), Some(other_float)) =
-                (self.value.to_bigint(), other_int.to_f64())
-            {
-                self.value >= other_float && self_int >= *other_int
-            } else {
-                // certainly not equal, forward to inner_gt_int
-                inner_gt_int(self.value, other_int)
-            };
-
-            vm.ctx.new_bool(result)
-        } else {
-            vm.ctx.not_implemented()
-        }
+    fn ge(&self, other: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        self.cmp(
+            other,
+            |a, b| a >= b,
+            |a, b| {
+                if let (Some(a_int), Some(b_float)) = (a.to_bigint(), b.to_f64()) {
+                    a >= b_float && a_int >= *b
+                } else {
+                    inner_gt_int(a, b)
+                }
+            },
+            vm,
+        )
     }
 
     #[pymethod(name = "__abs__")]
     fn abs(&self, _vm: &VirtualMachine) -> f64 {
         self.value.abs()
     }
 
-    #[pymethod(name = "__add__")]
-    fn add(&self, other: PyObjectRef, vm: &VirtualMachine) -> PyResult {
+    #[inline]
+    fn op<F>(&self, other: PyObjectRef, float_op: F, vm: &VirtualMachine) -> PyResult
```

Comment:
Instead of a single `op` function which works always, it might make sense to create several helper functions for specialized purposes, such as `binary_operator` for +,-,/ and *. The `divmod` operation requires the most generic form, `generic_binary_operation`.

Reply:
I made `simple_op` which returns `PyResult<f64>` and `tuple_op` which returns `PyResult<(f64, f64)>`. Please check if you like this change.